ml/backend/ggml/quant.go

```diff
@@ -0,0 +1,518 @@
+				new_type = C.GGML_TYPE_Q3_K
+			} else if ftype == fsggml.FileTypeQ3_K_M {
+				new_type = C.GGML_TYPE_Q4_K
+			} else if ftype == fsggml.FileTypeQ3_K_L {
+				new_type = C.GGML_TYPE_Q5_K
+			}
+		}
+	} else if strings.Contains(name, "attn_qkv.weight") {
+		if ftype == fsggml.FileTypeQ3_K_M || ftype == fsggml.FileTypeQ3_K_L {
+			new_type = C.GGML_TYPE_Q4_K
+		} else if ftype == fsggml.FileTypeQ4_K_M {
+			new_type = C.GGML_TYPE_Q5_K
+		} else if ftype == fsggml.FileTypeQ5_K_M {
+			new_type = C.GGML_TYPE_Q6_K
+		}
+	}
+
+	if C.ggml_is_quantized(new_type) == C._Bool(true) {
+		nx := shape[0]
+		ny := uint64(1)
+		if len(shape) > 1 {
+			ny = shape[1]
+		}
+		qk_k := uint64(C.ggml_blck_size(new_type))
+		if nx%qk_k != 0 {
+			slog.Warn(fmt.Sprintf("tensor cols %d x %d are not divisible by %d, required for %s.  Falling back to quantization %s", nx, ny, qk_k, C.GoString(C.ggml_type_name(new_type)), C.GoString(C.ggml_type_name(C.GGML_TYPE_F16))))
+			new_type = C.GGML_TYPE_F16
+		}
+	}
+	return new_type
+}
+
+// convertToF32 converts (dequantizes) the raw data to F32 so we can then quantize it
+func convertToF32(data []byte, dtype uint32, nelements uint64) []float32 {
+	f32_data := make([]float32, nelements)
+	elems := C.int64_t(nelements)
+	switch dtype {
+	case C.GGML_TYPE_F16:
+		C.ggml_fp16_to_fp32_row((*C.uint16_t)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	case C.GGML_TYPE_Q4_0:
+		C.dequantize_row_q4_0((*C.block_q4_0)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	case C.GGML_TYPE_Q4_1:
+		C.dequantize_row_q4_1((*C.block_q4_1)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	case C.GGML_TYPE_Q5_0:
+		C.dequantize_row_q5_0((*C.block_q5_0)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	case C.GGML_TYPE_Q5_1:
+		C.dequantize_row_q5_1((*C.block_q5_1)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	case C.GGML_TYPE_Q8_0:
+		C.dequantize_row_q8_0((*C.block_q8_0)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	case C.GGML_TYPE_Q2_K:
+		C.dequantize_row_q2_K((*C.block_q2_K)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	case C.GGML_TYPE_Q3_K:
+		C.dequantize_row_q3_K((*C.block_q3_K)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	case C.GGML_TYPE_Q4_K:
+		C.dequantize_row_q4_K((*C.block_q4_K)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	case C.GGML_TYPE_Q5_K:
+		C.dequantize_row_q5_K((*C.block_q5_K)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	case C.GGML_TYPE_Q6_K:
+		C.dequantize_row_q6_K((*C.block_q6_K)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	case C.GGML_TYPE_BF16:
+		C.ggml_bf16_to_fp32_row((*C.ggml_bf16_t)(unsafe.Pointer(&data[0])), (*C.float)(&f32_data[0]), elems)
+	default:
+		panic("unsupported quantization format")
+	}
+	return f32_data
+}
+
+func QuantizeModel(ctx context.Context, r *os.File, quantizeType string, fname_out string, fn func(progress float32)) error {
```

Comment:
This seems like the wrong abstraction since it couples the entire quantization to this package. Instead, I was picturing this living somewhere like `server/quantize.go` where it would be responsible for high level business logic. It would then call `ggml.Quantize` for each tensor which will perform the actual transformations

Reply:
ok